### PR TITLE
[InstCombine] Infer nsw/nuw for trunc

### PR DESCRIPTION
**Pull request by dtcxzyw**
This patch adds support for inferring trunc's nsw/nuw flags.

Alive2 support for these flags has landed as https://github.com/AliveToolkit/alive2/pull/1024. But alive2.llvm.org hasn't updated to the trunk. So I cannot provide the proof.

### CODE REVIEW

**Comment by dtcxzyw** [comment]
Compile-time impact: https://llvm-compile-time-tracker.com/compare.php?from=869797daca38941e0af3bcd8ae5300bcebf7b1a9&to=f4c5ea82ac929ba2283559d5970612977fe6845b&stat=instructions:u

File: llvm/lib/Transforms/InstCombine/InstCombineCasts.cpp [hunk 1]
```diff
@@ -897,7 +897,20 @@ Instruction *InstCombinerImpl::visitTrunc(TruncInst &Trunc) {
...
+      ComputeMaxSignificantBits(Src, /*Depth=*/0, &Trunc) <= DestWidth) {
+    Trunc.setHasNoSignedWrap(true);
+    Changed = true;
+  }
+  if (!Trunc.hasNoUnsignedWrap() &&
+      MaskedValueIsZero(Src, APInt::getBitsSetFrom(SrcWidth, DestWidth),
+                        /*Depth=*/0, &Trunc)) {
+    Trunc.setHasNoUnsignedWrap(true);
+    Changed = true;
+  }
```
Comment (goldsteinn) [inline llvm/lib/Transforms/InstCombine/InstCombineCasts.cpp:911]: From an efficiency POV, I think you should probably startout with a `computeKnownBits` on src and see if you get lucky about leading ones / zeros. Then if you don't, do a call to `computeMaxSignificantBits`.

**Comment by goldsteinn** [comment]
Not an issue now but I think we should look into updating `computeKnownBitsFromContext` to look use `trunc nsw/nuw` uses of `X`. If the use is dominating / noundef, we can infer bits about `X`.
Not really an issue if we only use `KnownBits` to infer `nuw`/`nsw`, but once we add support in `SCEV`/`CVP`/`SCCP`/etc.. it may be useful.

Guess same is true for `zext nneg` now (and could probably also do `sub nuw X, nonzero_Y` now in `isKnownNonNullFromDominatingCondition`).

**Comment by goldsteinn** [comment]
LGTM.

File: llvm/lib/Transforms/InstCombine/InstCombineCasts.cpp [hunk 2]
```diff
@@ -897,7 +897,20 @@ Instruction *InstCombinerImpl::visitTrunc(TruncInst &Trunc) {
...
+  if (!Trunc.hasNoSignedWrap() &&
+      ComputeMaxSignificantBits(Src, /*Depth=*/0, &Trunc) <= DestWidth) {
+    Trunc.setHasNoSignedWrap(true);
+    Changed = true;
+  }
+  if (!Trunc.hasNoUnsignedWrap() &&
+      MaskedValueIsZero(Src, APInt::getBitsSetFrom(SrcWidth, DestWidth),
+                        /*Depth=*/0, &Trunc)) {
+    Trunc.setHasNoUnsignedWrap(true);
+    Changed = true;
```
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineCasts.cpp:910]: Can we infer this in SimplifyDemanded instead, where we already have the KnownBits of the trunc arg?

Comment (dtcxzyw) [reply llvm/lib/Transforms/InstCombine/InstCombineCasts.cpp:910]: We cannot infer nsw flags from KnownBits (e.g., `trunc (ashr i64 X, 32) to i32`). BTW we never set poison-generating flags in `SimplifyDemanded`.

Comment (nikic) [reply llvm/lib/Transforms/InstCombine/InstCombineCasts.cpp:910]: We can't infer nsw, but we can infer nuw. Do you see any reason why doing this in SimplifyDemanded would be problematic?

Comment (dtcxzyw) [reply llvm/lib/Transforms/InstCombine/InstCombineCasts.cpp:910]: > We can't infer nsw, but we can infer nuw.

I prefer to infer both flags here, then we may reuse KnownBits in further patches.

> Do you see any reason why doing this in SimplifyDemanded would be problematic?

`SimplifyDemanded` is context-sensitive. IMO it is not the right place to infer flags.

Comment (dtcxzyw) [reply llvm/lib/Transforms/InstCombine/InstCombineCasts.cpp:910]: Ping @nikic

Comment (nikic) [reply llvm/lib/Transforms/InstCombine/InstCombineCasts.cpp:910]: Okay, I won't insist on this.

**Review by nikic — APPROVED** [review]
LGTM

**Pull request merged**